{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "94acffde-a523-4bc6-9fd7-d6c0487fbe02",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/abc.py:106: FutureWarning: xarray subclass CrAnData should explicitly define __slots__\n",
      "  cls = super().__new__(mcls, name, bases, namespace, **kwargs)\n",
      "100%|██████████| 2/2 [00:00<00:00, 318.26it/s]\n",
      "\u001b[32m2025-03-22 01:07:01.957\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mcrandata.chrom_io\u001b[0m:\u001b[36mimport_bigwigs\u001b[0m:\u001b[36m308\u001b[0m - \u001b[1mExtracting values from 2 bigWig files...\u001b[0m\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/core/array.py:3947: UserWarning: The dtype `StringDType()` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  result = await AsyncArray._create_v3(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/core/array.py:3947: UserWarning: The dtype `StringDType()` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  result = await AsyncArray._create_v3(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/core/array.py:3947: UserWarning: The dtype `StringDType()` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  result = await AsyncArray._create_v3(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/api/asynchronous.py:197: UserWarning: Consolidated metadata is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  warnings.warn(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/core/array.py:3947: UserWarning: The dtype `StringDType()` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  result = await AsyncArray._create_v3(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/api/asynchronous.py:197: UserWarning: Consolidated metadata is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  warnings.warn(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "100%|██████████| 2/2 [00:00<00:00, 59.78it/s]\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/core/group.py:2821: UserWarning: Object at var-_-chrom is not recognized as a component of a Zarr hierarchy.\n",
      "  warnings.warn(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/core/group.py:2821: UserWarning: Object at var-_-end is not recognized as a component of a Zarr hierarchy.\n",
      "  warnings.warn(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/core/group.py:2821: UserWarning: Object at var-_-index is not recognized as a component of a Zarr hierarchy.\n",
      "  warnings.warn(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/core/group.py:2821: UserWarning: Object at var-_-chunk_index is not recognized as a component of a Zarr hierarchy.\n",
      "  warnings.warn(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/core/group.py:2821: UserWarning: Object at var-_-start is not recognized as a component of a Zarr hierarchy.\n",
      "  warnings.warn(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/core/group.py:2821: UserWarning: Object at obs-_-file_path is not recognized as a component of a Zarr hierarchy.\n",
      "  warnings.warn(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/core/group.py:2821: UserWarning: Object at obs-_-index is not recognized as a component of a Zarr hierarchy.\n",
      "  warnings.warn(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/core/group.py:2821: UserWarning: Object at X is not recognized as a component of a Zarr hierarchy.\n",
      "  warnings.warn(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/core/group.py:2821: UserWarning: Object at obs is not recognized as a component of a Zarr hierarchy.\n",
      "  warnings.warn(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/core/group.py:2821: UserWarning: Object at var is not recognized as a component of a Zarr hierarchy.\n",
      "  warnings.warn(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/core/group.py:2821: UserWarning: Object at seq_bins is not recognized as a component of a Zarr hierarchy.\n",
      "  warnings.warn(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/core/array.py:3947: UserWarning: The dtype `StringDType()` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  result = await AsyncArray._create_v3(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/core/array.py:3947: UserWarning: The dtype `StringDType()` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  result = await AsyncArray._create_v3(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/core/array.py:3947: UserWarning: The dtype `StringDType()` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  result = await AsyncArray._create_v3(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/core/array.py:3947: UserWarning: The dtype `StringDType()` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  result = await AsyncArray._create_v3(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/api/asynchronous.py:197: UserWarning: Consolidated metadata is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  warnings.warn(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/api/asynchronous.py:197: UserWarning: Consolidated metadata is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  warnings.warn(\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/grandada/crandata/_module.py:88: FutureWarning: The return type of `Dataset.dims` will be changed to return a set of dimension names in future, in order to be more consistent with `DataArray.dims`. To access a mapping from dimension names to lengths, please use `Dataset.sizes`.\n",
      "  dim_dict = dict(self.adata.dims)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/xbatcher/generators.py:112: UserWarning: The following dimensions were included in both ``input_dims`` and ``batch_dims``. Since ``concat_input_dims`` is ``False``, these dimensions will not impact batch generation: {'var': 3}\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded CrAnData:\n",
      "CrAnData object\n",
      "Array names: ['obs-_-index', 'var-_-chunk_index', 'X', 'var-_-end', 'var-_-chrom', 'obs-_-file_path', 'var-_-start', 'var-_-index', 'sequences']\n",
      "Coordinates: ['obs', 'var', 'seq_bins']\n",
      "\n",
      "\n",
      "Iterating over a couple of training batches from MetaCrAnDataModule:\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1it [00:00,  1.13it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samples [CrAnData object\n",
      "Array names: ['obs-_-index', 'var-_-chunk_index', 'X', 'var-_-end', 'var-_-chrom', 'obs-_-file_path', 'var-_-start', 'var-_-index', 'sequences', 'var-_-split']\n",
      "Coordinates: ['obs', 'var', 'seq_bins']\n",
      ", CrAnData object\n",
      "Array names: ['obs-_-index', 'var-_-chunk_index', 'X', 'var-_-end', 'var-_-chrom', 'obs-_-file_path', 'var-_-start', 'var-_-index', 'sequences', 'var-_-split']\n",
      "Coordinates: ['obs', 'var', 'seq_bins']\n",
      ", CrAnData object\n",
      "Array names: ['obs-_-index', 'var-_-chunk_index', 'X', 'var-_-end', 'var-_-chrom', 'obs-_-file_path', 'var-_-start', 'var-_-index', 'sequences', 'var-_-split']\n",
      "Coordinates: ['obs', 'var', 'seq_bins']\n",
      "]\n",
      "CrAnData object\n",
      "Array names: ['obs-_-index', 'var-_-chunk_index', 'X', 'var-_-end', 'var-_-chrom', 'obs-_-file_path', 'var-_-start', 'var-_-index', 'sequences', 'var-_-split']\n",
      "Coordinates: ['obs', 'var', 'seq_bins']\n",
      "\n",
      "\n",
      "Meta Batch 0:\n",
      "  obs-_-index: shape (2,)\n",
      "  var-_-chunk_index: shape (3,)\n",
      "  X: shape (2, 3, 100)\n",
      "  var-_-end: shape (3,)\n",
      "  var-_-chrom: shape (3,)\n",
      "  obs-_-file_path: shape (2,)\n",
      "  var-_-start: shape (3,)\n",
      "  var-_-index: shape (3,)\n",
      "  sequences: shape (3, 80, 4)\n",
      "  var-_-split: shape (3,)\n",
      "samples [CrAnData object\n",
      "Array names: ['obs-_-index', 'var-_-chunk_index', 'X', 'var-_-end', 'var-_-chrom', 'obs-_-file_path', 'var-_-start', 'var-_-index', 'sequences', 'var-_-split']\n",
      "Coordinates: ['obs', 'var', 'seq_bins']\n",
      ", CrAnData object\n",
      "Array names: ['obs-_-index', 'var-_-chunk_index', 'X', 'var-_-end', 'var-_-chrom', 'obs-_-file_path', 'var-_-start', 'var-_-index', 'sequences', 'var-_-split']\n",
      "Coordinates: ['obs', 'var', 'seq_bins']\n",
      ", CrAnData object\n",
      "Array names: ['obs-_-index', 'var-_-chunk_index', 'X', 'var-_-end', 'var-_-chrom', 'obs-_-file_path', 'var-_-start', 'var-_-index', 'sequences', 'var-_-split']\n",
      "Coordinates: ['obs', 'var', 'seq_bins']\n",
      "]\n",
      "CrAnData object\n",
      "Array names: ['obs-_-index', 'var-_-chunk_index', 'X', 'var-_-end', 'var-_-chrom', 'obs-_-file_path', 'var-_-start', 'var-_-index', 'sequences', 'var-_-split']\n",
      "Coordinates: ['obs', 'var', 'seq_bins']\n",
      "\n",
      "\n",
      "Meta Batch 1:\n",
      "  obs-_-index: shape (2,)\n",
      "  var-_-chunk_index: shape (3,)\n",
      "  X: shape (2, 3, 100)\n",
      "  var-_-end: shape (3,)\n",
      "  var-_-chrom: shape (3,)\n",
      "  obs-_-file_path: shape (2,)\n",
      "  var-_-start: shape (3,)\n",
      "  var-_-index: shape (3,)\n",
      "  sequences: shape (3, 80, 4)\n",
      "  var-_-split: shape (3,)\n",
      "\n",
      "Temporary directory contents:\n",
      "['beds', 'bigwigs', 'chrom.sizes', 'consensus.bed', 'chrom_data.zarr', 'chr1.fa', 'chr1.fa.fai']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import tempfile\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pyBigWig\n",
    "import copy\n",
    "import xarray as xr\n",
    "import tqdm\n",
    "\n",
    "# Import our new module system and utilities.\n",
    "from crandata import CrAnDataModule, MetaCrAnDataModule, CrAnData\n",
    "from crandata.chrom_io import import_bigwigs\n",
    "from crandata.seq_io import add_genome_sequences_to_crandata, DNATransform\n",
    "\n",
    "# Create temporary directories for synthetic data.\n",
    "temp_dir = tempfile.TemporaryDirectory()\n",
    "base_dir = Path(temp_dir.name)\n",
    "beds_dir = base_dir / \"beds\"\n",
    "bigwigs_dir = base_dir / \"bigwigs\"\n",
    "beds_dir.mkdir(exist_ok=True)\n",
    "bigwigs_dir.mkdir(exist_ok=True)\n",
    "\n",
    "# Create a chromsizes file.\n",
    "chromsizes_file = base_dir / \"chrom.sizes\"\n",
    "with open(chromsizes_file, \"w\") as f:\n",
    "    f.write(\"chr1\\t1000\\n\")\n",
    "\n",
    "# Create two BED files (simulate two different classes).\n",
    "bed_data_A = pd.DataFrame({\n",
    "    0: [\"chr1\", \"chr1\"],\n",
    "    1: [100, 300],\n",
    "    2: [200, 400]\n",
    "})\n",
    "bed_data_B = pd.DataFrame({\n",
    "    0: [\"chr1\", \"chr1\"],\n",
    "    1: [150, 350],\n",
    "    2: [250, 450]\n",
    "})\n",
    "bed_file_A = beds_dir / \"ClassA.bed\"\n",
    "bed_file_B = beds_dir / \"ClassB.bed\"\n",
    "bed_data_A.to_csv(bed_file_A, sep=\"\\t\", header=False, index=False)\n",
    "bed_data_B.to_csv(bed_file_B, sep=\"\\t\", header=False, index=False)\n",
    "\n",
    "# Create a consensus BED file.\n",
    "consensus = pd.DataFrame({\n",
    "    0: [\"chr1\", \"chr1\", \"chr1\"],\n",
    "    1: [100, 300, 350],\n",
    "    2: [200, 400, 450]\n",
    "})\n",
    "consensus_file = base_dir / \"consensus.bed\"\n",
    "consensus.to_csv(consensus_file, sep=\"\\t\", header=False, index=False)\n",
    "\n",
    "# Create two bigWig files.\n",
    "bigwig_file1 = bigwigs_dir / \"test.bw\"\n",
    "bw1 = pyBigWig.open(str(bigwig_file1), \"w\")\n",
    "bw1.addHeader([(\"chr1\", 1000)])\n",
    "bw1.addEntries(chroms=[\"chr1\"], starts=[0], ends=[1000], values=[5.0])\n",
    "bw1.close()\n",
    "\n",
    "bigwig_file2 = bigwigs_dir / \"test2.bw\"\n",
    "bw2 = pyBigWig.open(str(bigwig_file2), \"w\")\n",
    "bw2.addHeader([(\"chr1\", 1000)])\n",
    "bw2.addEntries(chroms=[\"chr1\"], starts=[0], ends=[1000], values=[4.0])\n",
    "bw2.close()\n",
    "\n",
    "# Set extraction parameters.\n",
    "target_region_width = 100\n",
    "backed_path = base_dir / \"chrom_data.zarr\"\n",
    "\n",
    "# Create the CrAnData object from bigWig files and consensus regions.\n",
    "adata = import_bigwigs(\n",
    "    bigwigs_folder=str(bigwigs_dir),\n",
    "    regions_file=str(consensus_file),\n",
    "    backed_path=str(backed_path),\n",
    "    target_region_width=target_region_width,\n",
    "    chromsizes_file=str(chromsizes_file),\n",
    ")\n",
    "\n",
    "# Create a dummy FASTA file for a genome.\n",
    "fasta_file = base_dir / \"chr1.fa\"\n",
    "with open(fasta_file, \"w\") as f:\n",
    "    f.write(\">chr1\\n\")\n",
    "    f.write(\"A\" * 1000 + \"\\n\")\n",
    "\n",
    "# Create a Genome object.\n",
    "from crandata._genome import Genome\n",
    "dummy_genome = Genome(str(fasta_file), chrom_sizes=str(chromsizes_file))\n",
    "\n",
    "# Add sequences to the CrAnData using the provided seq_io utility.\n",
    "# Here we use the consensus regions as our ranges.\n",
    "consensus.columns = ['chrom', 'start', 'end']\n",
    "adata = add_genome_sequences_to_crandata(adata, consensus, dummy_genome)\n",
    "\n",
    "# Write the CrAnData object to disk and then reload it to ensure sequences are out-of-memory.\n",
    "adata.to_zarr(str(backed_path),mode='a')\n",
    "adata_loaded = CrAnData.open_zarr(str(backed_path))\n",
    "print(\"Loaded CrAnData:\")\n",
    "print(adata_loaded)\n",
    "\n",
    "# Create two copies to simulate two datasets (e.g. two species), and add a \"split\" column in var metadata.\n",
    "adata1 = copy.deepcopy(adata_loaded)\n",
    "adata2 = copy.deepcopy(adata_loaded)\n",
    "adata1[\"var-_-split\"] = xr.DataArray(np.full(adata1.sizes[\"var\"], \"train\"), dims=[\"var\"])\n",
    "adata2[\"var-_-split\"] = xr.DataArray(np.full(adata2.sizes[\"var\"], \"train\"), dims=[\"var\"])\n",
    "\n",
    "# Create a DNATransform instance.\n",
    "transform = DNATransform(out_len=80, random_rc=True, max_shift=5)\n",
    "\n",
    "# Instantiate the MetaCrAnDataModule with the two datasets.\n",
    "# Note: The batch_size is now 3, matching the number of consensus regions (var dimension).\n",
    "meta_module = MetaCrAnDataModule(\n",
    "    adatas=[adata1, adata2],\n",
    "    batch_size=3,        # adjust batch size to not exceed var length (3)\n",
    "    shuffle=True,\n",
    "    dnatransform=transform,\n",
    "    epoch_size=10\n",
    ")\n",
    "\n",
    "# Setup each underlying module for the \"train\" stage.\n",
    "for mod in meta_module.modules:\n",
    "    mod.setup(state=\"train\")\n",
    "\n",
    "# Retrieve the training dataloader from the meta module and iterate over a couple of batches.\n",
    "meta_train_dl = meta_module.train_dataloader\n",
    "print(\"\\nIterating over a couple of training batches from MetaCrAnDataModule:\")\n",
    "for i, batch in enumerate(tqdm.tqdm(meta_train_dl)):\n",
    "    print(batch)\n",
    "    print(f\"\\nMeta Batch {i}:\")\n",
    "    for key, tensor in batch.items():\n",
    "        print(f\"  {key}: shape {tensor.shape}\")\n",
    "    if i >= 1:\n",
    "        break\n",
    "\n",
    "print(\"\\nTemporary directory contents:\")\n",
    "print(os.listdir(base_dir))\n",
    "temp_dir.cleanup()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2cb0a57-e333-4776-9786-e3eeaf1ba370",
   "metadata": {},
   "outputs": [],
   "source": [
    "sdfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ac6455b-1f4e-4b65-8ea8-e15602c43ff5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Should the fill in _extract_values_from_bigwig actually be 0? Can we filter var where all is 0/nan without loading everything into memory?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5f70292f-f088-4e9d-a127-4736bb591964",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/abc.py:106: FutureWarning: xarray subclass CrAnData should explicitly define __slots__\n",
      "  cls = super().__new__(mcls, name, bases, namespace, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "import crandata\n",
    "import xarray as xr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import crested\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b454e2de-689f-432e-9621-4d88f73e3b7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "genomes = {}\n",
    "beds = {}\n",
    "chromsizes_files = {}\n",
    "bed_files = {}\n",
    "species = ['mouse','human','macaque']\n",
    "\n",
    "WINDOW_SIZE = 2114\n",
    "OFFSET = WINDOW_SIZE // 2  # e.g., 50% overlap\n",
    "N_THRESHOLD = 0.3\n",
    "n_bins = WINDOW_SIZE//50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "392bdeca-411d-47ea-a846-6c906784b46d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Calculating N content: 100%|██████████| 2583507/2583507 [02:37<00:00, 16430.01it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered bins:\n",
      "              chrom    start      end    prop_n\n",
      "0              chr1  2999767  3001881  0.110638\n",
      "1              chr1  3002938  3005052  0.085579\n",
      "2              chr1  3003995  3006109  0.000000\n",
      "3              chr1  3005052  3007166  0.000000\n",
      "4              chr1  3006109  3008223  0.000000\n",
      "...             ...      ...      ...       ...\n",
      "2509462  JH584292.1     8457    10571  0.000000\n",
      "2509463  JH584292.1     9514    11628  0.000000\n",
      "2509464  JH584292.1    10571    12685  0.000000\n",
      "2509465  JH584292.1    11628    13742  0.000000\n",
      "2509466  JH584292.1    12685    14799  0.000000\n",
      "\n",
      "[2509467 rows x 4 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Calculating N content: 100%|██████████| 2932321/2932321 [03:03<00:00, 15951.14it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered bins:\n",
      "              chrom  start    end   prop_n\n",
      "0              chr1   9514  11628  0.23026\n",
      "1              chr1  10571  12685  0.00000\n",
      "2              chr1  11628  13742  0.00000\n",
      "3              chr1  12685  14799  0.00000\n",
      "4              chr1  13742  15856  0.00000\n",
      "...             ...    ...    ...      ...\n",
      "2786513  KI270518.1      1   2115  0.00000\n",
      "2786514  KI270530.1      1   2115  0.00000\n",
      "2786515  KI270304.1      1   2115  0.00000\n",
      "2786516  KI270418.1      1   2115  0.00000\n",
      "2786517  KI270424.1      1   2115  0.00000\n",
      "\n",
      "[2786518 rows x 4 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Calculating N content: 100%|██████████| 2806701/2806701 [02:50<00:00, 16497.35it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered bins:\n",
      "               chrom  start    end  prop_n\n",
      "0        NC_041754.1      1   2115     0.0\n",
      "1        NC_041754.1   1058   3172     0.0\n",
      "2        NC_041754.1   2115   4229     0.0\n",
      "3        NC_041754.1   3172   5286     0.0\n",
      "4        NC_041754.1   4229   6343     0.0\n",
      "...              ...    ...    ...     ...\n",
      "2773980  NC_005943.1   9514  11628     0.0\n",
      "2773981  NC_005943.1  10571  12685     0.0\n",
      "2773982  NC_005943.1  11628  13742     0.0\n",
      "2773983  NC_005943.1  12685  14799     0.0\n",
      "2773984  NC_005943.1  13742  15856     0.0\n",
      "\n",
      "[2773985 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "for s in species:\n",
    "    genome_path = '/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/genome/onehots/'+s\n",
    "    fasta_file = os.path.join(genome_path,s+'.fa')\n",
    "    chrom_sizes = os.path.join(genome_path,s+'.fa.sizes')\n",
    "    annotation_gtf_file = os.path.join(genome_path,s+'.annotation.gtf')\n",
    "    chromsizes_files[s] = chrom_sizes\n",
    "    genome = crandata.Genome(fasta_file, chrom_sizes, annotation_gtf_file)\n",
    "    genomes[s] = genome\n",
    "    # Set parameters for binning.\n",
    "    \n",
    "    # Optionally specify an output path for the BED file.\n",
    "    OUTPUT_BED = os.path.join(genome_path, \"binned_genome.bed\")\n",
    "    bed_files[s] = OUTPUT_BED\n",
    "    # Generate bins and optionally write to disk.\n",
    "    binned_df = crandata.bin_genome(genome, WINDOW_SIZE, OFFSET, n_threshold=N_THRESHOLD, output_path=OUTPUT_BED).reset_index(drop=True)\n",
    "    print(\"Filtered bins:\")\n",
    "    print(binned_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b9a7ce09-e4f1-4c16-bdc3-e8f390dde2ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "import importlib\n",
    "crandata.chrom_io = importlib.reload(crandata.chrom_io)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "10492d8e-e22d-48a3-8139-6d5aadc02d7f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/zarr/codecs/vlen_utf8.py:44: UserWarning: The codec `vlen-utf8` is currently not part in the Zarr format 3 specification. It may not be supported by other zarr implementations and may change in the future.\n",
      "  return cls(**configuration_parsed)\n"
     ]
    }
   ],
   "source": [
    "adatas = {}\n",
    "\n",
    "for s in species:\n",
    "    bigwigs_dir = os.path.join('/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/SpinalCord/manuscript/ATAC',s,'Group_bigwig')\n",
    "    # adata = crandata.chrom_io.import_bigwigs(\n",
    "    #     bigwigs_folder=bigwigs_dir,\n",
    "    #     regions_file=bed_files[s],\n",
    "    #     backed_path='/home/matthew.schmitz/Matthew/'+s+'_spc_test.zarr',\n",
    "    #     target_region_width=WINDOW_SIZE,\n",
    "    #     chromsizes_file=chromsizes_files[s],\n",
    "    #     target = 'raw',\n",
    "    #     n_bins=n_bins\n",
    "    # )\n",
    "    # adatas[s] = adata\n",
    "    adatas[s] = crandata.crandata.CrAnData.open_zarr('/home/matthew.schmitz/Matthew/'+s+'_spc_test.zarr')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a0cd7c04-0d92-4300-b098-e3f0431b7071",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'mouse': CrAnData object\n",
       " Array names: ['obs-_-index', 'var-_-chunk_index', 'X', 'obs-_-file_path', 'var-_-chrom', 'var-_-end', 'var-_-start', 'var-_-index']\n",
       " Coordinates: ['obs', 'var', 'seq_bins'],\n",
       " 'human': CrAnData object\n",
       " Array names: ['obs-_-index', 'obs-_-file_path', 'X', 'var-_-chunk_index', 'var-_-start', 'var-_-end', 'var-_-chrom', 'var-_-index']\n",
       " Coordinates: ['obs', 'seq_bins', 'var'],\n",
       " 'macaque': CrAnData object\n",
       " Array names: ['obs-_-index', 'X', 'var-_-start', 'var-_-chunk_index', 'var-_-chrom', 'obs-_-file_path', 'var-_-end', 'var-_-index']\n",
       " Coordinates: ['obs', 'seq_bins', 'var']}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adatas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7fa88d3-f313-466e-858e-9e8a20af3b04",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import numpy as np\n",
    "# adatas['mouse'].uns['chunk_size'] = 512\n",
    "# adatas['human'].uns['chunk_size'] = 512\n",
    "# adatas['macaque'].uns['chunk_size'] = 512\n",
    "# adatas['mouse'].var[\"chunk_index\"] = np.arange(adatas['mouse'].var.shape[0]) // 512\n",
    "# adatas['human'].var[\"chunk_index\"] = np.arange(adatas['human'].var.shape[0]) // 512\n",
    "# adatas['macaque'].var[\"chunk_index\"] = np.arange(adatas['macaque'].var.shape[0]) // 512\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b667671b-3508-4daf-a706-92651c186854",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'CrAnData' object has no attribute 'var_names'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m s \u001b[38;5;129;01min\u001b[39;00m adatas\u001b[38;5;241m.\u001b[39mkeys():\n\u001b[0;32m----> 2\u001b[0m     \u001b[43mcrested\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtrain_val_test_split\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m      3\u001b[0m \u001b[43m        \u001b[49m\u001b[43madatas\u001b[49m\u001b[43m[\u001b[49m\u001b[43ms\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstrategy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mregion\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mval_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0.1\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtest_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0.1\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrandom_state\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m42\u001b[39;49m\n\u001b[1;32m      4\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/crested/pp/_split.py:279\u001b[0m, in \u001b[0;36mtrain_val_test_split\u001b[0;34m(adata, strategy, val_size, test_size, val_chroms, test_chroms, shuffle, random_state)\u001b[0m\n\u001b[1;32m    274\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    275\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mIf `strategy` is \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mchr\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m, `val_chroms` and `test_chroms` should be provided.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    276\u001b[0m         )\n\u001b[1;32m    278\u001b[0m \u001b[38;5;66;03m# Split\u001b[39;00m\n\u001b[0;32m--> 279\u001b[0m regions \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(\u001b[43madata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvar_names\u001b[49m)\n\u001b[1;32m    281\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m strategy \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mregion\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    282\u001b[0m     split \u001b[38;5;241m=\u001b[39m _split_by_regions(regions, val_size, test_size, shuffle, random_state)\n",
      "File \u001b[0;32m~/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/xarray/core/common.py:305\u001b[0m, in \u001b[0;36mAttrAccessMixin.__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m    303\u001b[0m         \u001b[38;5;28;01mwith\u001b[39;00m suppress(\u001b[38;5;167;01mKeyError\u001b[39;00m):\n\u001b[1;32m    304\u001b[0m             \u001b[38;5;28;01mreturn\u001b[39;00m source[name]\n\u001b[0;32m--> 305\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mAttributeError\u001b[39;00m(\n\u001b[1;32m    306\u001b[0m     \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(\u001b[38;5;28mself\u001b[39m)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m object has no attribute \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mname\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    307\u001b[0m )\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'CrAnData' object has no attribute 'var_names'"
     ]
    }
   ],
   "source": [
    "for s in adatas.keys():\n",
    "    crested.pp.train_val_test_split(\n",
    "        adatas[s], strategy=\"region\", val_size=0.1, test_size=0.1, random_state=42\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c9aaf99e-b31e-4f93-a15d-e6ca8aa65780",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "MetaCrAnDataModule.__init__() got an unexpected keyword argument 'in_memory'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m meta_module \u001b[38;5;241m=\u001b[39m \u001b[43mcrandata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mMetaCrAnDataModule\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m      2\u001b[0m \u001b[43m    \u001b[49m\u001b[43madatas\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mlist\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43madatas\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalues\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      3\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;66;43;03m# genomes=list(genomes.values()),\u001b[39;49;00m\n\u001b[1;32m      4\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;66;43;03m# data_sources={'y': 'X'},\u001b[39;49;00m\n\u001b[1;32m      5\u001b[0m \u001b[43m    \u001b[49m\u001b[43min_memory\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m      6\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrandom_reverse_complement\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m      7\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmax_stochastic_shift\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m10\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      8\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdeterministic_shift\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m      9\u001b[0m \u001b[43m    \u001b[49m\u001b[43mshuffle_obs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mobs_alignment\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mintersect\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     10\u001b[0m \u001b[43m    \u001b[49m\u001b[43mshuffle\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m     11\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbatch_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m32\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m    \u001b[49m\u001b[38;5;66;43;03m# small batch size for testing\u001b[39;49;00m\n\u001b[1;32m     12\u001b[0m \u001b[43m    \u001b[49m\u001b[43mepoch_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1000000\u001b[39;49m\u001b[43m    \u001b[49m\u001b[38;5;66;43;03m# small epoch size for quick testing\u001b[39;49;00m\n\u001b[1;32m     13\u001b[0m \u001b[43m)\u001b[49m\n\u001b[1;32m     15\u001b[0m \u001b[38;5;66;03m# Setup the meta module for the \"fit\" stage (train/val)\u001b[39;00m\n\u001b[1;32m     16\u001b[0m meta_module\u001b[38;5;241m.\u001b[39msetup(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfit\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mTypeError\u001b[0m: MetaCrAnDataModule.__init__() got an unexpected keyword argument 'in_memory'"
     ]
    }
   ],
   "source": [
    "meta_module = crandata.MetaCrAnDataModule(\n",
    "    adatas=list(adatas.values()),\n",
    "    # genomes=list(genomes.values()),\n",
    "    # data_sources={'y': 'X'},\n",
    "    in_memory=False,\n",
    "    random_reverse_complement=True,\n",
    "    max_stochastic_shift=10,\n",
    "    deterministic_shift=False,\n",
    "    shuffle_obs=False, obs_alignment = 'intersect',\n",
    "    shuffle=True,\n",
    "    batch_size=32,    # small batch size for testing\n",
    "    epoch_size=1000000    # small epoch size for quick testing\n",
    ")\n",
    "\n",
    "# Setup the meta module for the \"fit\" stage (train/val)\n",
    "meta_module.setup(\"fit\")\n",
    "\n",
    "# Retrieve the training dataloader from the meta module and iterate over a couple of batches.\n",
    "meta_train_dl = meta_module.train_dataloader\n",
    "\n",
    "print(\"\\nIterating over a couple of training batches from MetaAnnDataModule:\")\n",
    "for i, batch in enumerate(tqdm(meta_train_dl.data)):\n",
    "    print(f\"Meta Batch {i}:\")\n",
    "    for key, tensor in batch.items():\n",
    "        print(f\"  {key}: shape {tensor.shape}\")\n",
    "    if i == 5:\n",
    "        break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a52795a1-dccf-4a75-a9c8-e5ee4b416051",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bgen1 has 4 batches\n",
      "First batch from var1:\n",
      "<xarray.Dataset> Size: 10kB\n",
      "Dimensions:  (time: 5, sample: 150)\n",
      "Coordinates:\n",
      "  * sample   (sample) object 1kB MultiIndex\n",
      "  * lat      (sample) int64 1kB 0 0 0 0 0 0 0 0 0 0 0 ... 9 9 9 9 9 9 9 9 9 9 9\n",
      "  * lon      (sample) int64 1kB 0 1 2 3 4 5 6 7 8 9 ... 5 6 7 8 9 10 11 12 13 14\n",
      "Dimensions without coordinates: time\n",
      "Data variables:\n",
      "    var1     (sample, time) float64 6kB 0.2816 0.607 0.9085 ... 0.815 0.3227\n",
      "\n",
      "bgen2 has 3 batches\n",
      "First batch from var2:\n",
      "<xarray.Dataset> Size: 3kB\n",
      "Dimensions:  (x: 10, y: 40)\n",
      "Dimensions without coordinates: x, y\n",
      "Data variables:\n",
      "    var2     (x, y) float64 3kB 0.4145 0.2682 0.2274 ... 0.9241 0.4045 0.5468\n",
      "\n",
      "bgen3 has 2 batches\n",
      "First batch from var5:\n",
      "<xarray.Dataset> Size: 448B\n",
      "Dimensions:  (time: 7, level: 8)\n",
      "Dimensions without coordinates: time, level\n",
      "Data variables:\n",
      "    var5     (time, level) float64 448B 0.6049 0.9134 0.6541 ... 0.5835 0.5764\n",
      "\n",
      "bgen4 (with overlap) has 2 batches\n",
      "First batch from var1 with overlap:\n",
      "<xarray.Dataset> Size: 108kB\n",
      "Dimensions:      (time_input: 5, sample: 1500)\n",
      "Coordinates:\n",
      "  * sample       (sample) object 12kB MultiIndex\n",
      "  * input_batch  (sample) int64 12kB 0 0 0 0 0 0 0 0 0 0 ... 9 9 9 9 9 9 9 9 9 9\n",
      "  * lat          (sample) int64 12kB 0 0 0 0 0 0 0 0 0 0 ... 9 9 9 9 9 9 9 9 9 9\n",
      "  * lon          (sample) int64 12kB 0 1 2 3 4 5 6 7 8 ... 7 8 9 10 11 12 13 14\n",
      "Dimensions without coordinates: time_input\n",
      "Data variables:\n",
      "    var1         (sample, time_input) float64 60kB 0.2816 0.607 ... 0.883 0.7146\n",
      "\n",
      "Iterating over batches in bgen1:\n",
      "Batch 0: var1 shape = (150, 5)\n",
      "Batch 1: var1 shape = (150, 5)\n",
      "Batch 2: var1 shape = (150, 5)\n",
      "Batch 3: var1 shape = (150, 5)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import xarray as xr\n",
    "import xbatcher\n",
    "\n",
    "# Create an xarray Dataset with 5 variables of various shapes and dimensions\n",
    "ds = xr.Dataset({\n",
    "    'var1': (('time', 'lat', 'lon'), np.random.rand(20, 10, 15)),      # e.g. climate data\n",
    "    'var2': (('x', 'y'), np.random.rand(30, 40)),                        # e.g. 2D image-like array\n",
    "    'var3': (('sample', 'feature'), np.random.rand(50, 5)),              # e.g. tabular features\n",
    "    'var4': (('channel', 'height', 'width'), np.random.rand(3, 32, 32)),   # e.g. multi-channel image\n",
    "    'var5': (('time', 'level'), np.random.rand(20, 8))                   # e.g. time-series with levels\n",
    "})\n",
    "\n",
    "# Example 1: Use xbatcher on var1 along the 'time' dimension\n",
    "# Each batch will contain 5 time points. (20/5 = 4 batches expected.)\n",
    "bgen1 = xbatcher.BatchGenerator(ds=ds[['var1']], input_dims={'time': 5})\n",
    "print(f'bgen1 has {len(bgen1)} batches')\n",
    "print(\"First batch from var1:\")\n",
    "print(bgen1[0])\n",
    "\n",
    "# Example 2: Use xbatcher on var2 along the 'x' dimension\n",
    "# Each batch will contain 10 x-points. (30/10 = 3 batches expected.)\n",
    "bgen2 = xbatcher.BatchGenerator(ds=ds[['var2']], input_dims={'x': 10})\n",
    "print(f'\\nbgen2 has {len(bgen2)} batches')\n",
    "print(\"First batch from var2:\")\n",
    "print(bgen2[0])\n",
    "\n",
    "# Example 3: Use xbatcher on var5 along the 'time' dimension\n",
    "# Here the chosen window size (7) does not evenly divide the 20 time points,\n",
    "# so the remainder will be discarded.\n",
    "bgen3 = xbatcher.BatchGenerator(ds=ds[['var5']], input_dims={'time': 7})\n",
    "print(f'\\nbgen3 has {len(bgen3)} batches')\n",
    "print(\"First batch from var5:\")\n",
    "print(bgen3[0])\n",
    "\n",
    "# Example 4: Batch generator for var1 with overlapping inputs\n",
    "# Each sample has 5 time points and each batch is 10 time points long,\n",
    "# with an overlap of 4 time points between consecutive samples.\n",
    "bgen4 = xbatcher.BatchGenerator(\n",
    "    ds=ds[['var1']],\n",
    "    input_dims={'time': 5},\n",
    "    batch_dims={'time': 10},\n",
    "    concat_input_dims=True,\n",
    "    input_overlap={'time': 4}\n",
    ")\n",
    "print(f'\\nbgen4 (with overlap) has {len(bgen4)} batches')\n",
    "print(\"First batch from var1 with overlap:\")\n",
    "print(bgen4[0])\n",
    "\n",
    "# Example 5: Iterate through batches in bgen1 and print the shape of var1 in each batch\n",
    "print(\"\\nIterating over batches in bgen1:\")\n",
    "for i, batch in enumerate(bgen1):\n",
    "    print(f\"Batch {i}: var1 shape = {batch.var1.shape}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de99d0b3-99e8-4a30-8cf4-b762a1247d32",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, batch in enumerate(tqdm(meta_train_dl.data)):\n",
    "    print(f\"Meta Batch {i}:\")\n",
    "    for key, tensor in batch.items():\n",
    "        print(f\"  {key}: shape {tensor.dtype}\")\n",
    "    if i == 5:\n",
    "        break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a11e2cb-b92b-4c02-bc40-886f67d2e8d5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import cProfile\n",
    "\n",
    "code = '''\n",
    "for i, batch in enumerate(meta_train_dl.data):\n",
    "    # print(f\"Meta Batch {i}:\")\n",
    "    # for key, tensor in batch.items():\n",
    "    #     print(f\"  {key}: shape {tensor.shape}\")\n",
    "    if i == 5:\n",
    "        break\n",
    "'''\n",
    "\n",
    "out = cProfile.run(code,sort=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e661681c-6d1e-4441-9878-c21d0505df51",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_architecture = crested.tl.zoo.simple_convnet(\n",
    "    seq_len=2114, num_classes=batch['y'].shape[1]\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6348c92d-cfce-4e39-8a7f-3df53b363b77",
   "metadata": {},
   "outputs": [],
   "source": [
    "import keras\n",
    "# Create your own configuration\n",
    "# I recommend trying this for peak regression with a weighted cosine mse log loss function\n",
    "optimizer = keras.optimizers.Adam(learning_rate=1e-5)\n",
    "loss = crested.tl.losses.CosineMSELogLoss(max_weight=100, multiplier=1)\n",
    "loss = crested.tl.losses.PoissonLoss()\n",
    "\n",
    "metrics = [\n",
    "    keras.metrics.MeanAbsoluteError(),\n",
    "    # keras.metrics.MeanSquaredError(),\n",
    "    # keras.metrics.CosineSimilarity(axis=1),\n",
    "    crested.tl.metrics.PearsonCorrelation(),\n",
    "    # crested.tl.metrics.ConcordanceCorrelationCoefficient(),\n",
    "    # crested.tl.metrics.PearsonCorrelationLog(),\n",
    "    # crested.tl.metrics.ZeroPenaltyMetric(),\n",
    "]\n",
    "\n",
    "alternative_config = crested.tl.TaskConfig(optimizer, loss, metrics)\n",
    "print(alternative_config)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80e5aeb3-5514-466c-b8c7-f04556666c35",
   "metadata": {},
   "outputs": [],
   "source": [
    "# initialize some lazy model parameters *yawn*\n",
    "model_architecture(batch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1b3716a-4a11-418e-9aad-6778019c0f97",
   "metadata": {},
   "outputs": [],
   "source": [
    "trainer = crested.tl.Crested(\n",
    "    data=meta_module,\n",
    "    model=model_architecture,\n",
    "    config=alternative_config,\n",
    "    project_name=\"mouse_biccn\",  # change to your liking\n",
    "    run_name=\"basemodel\",  # change to your liking\n",
    "    logger=None,  # or None, 'dvc', 'tensorboard'\n",
    "    seed=7,  # For reproducibility\n",
    ")\n",
    "# train the model\n",
    "trainer.fit(\n",
    "    epochs=60,\n",
    "    learning_rate_reduce_patience=3,\n",
    "    early_stopping_patience=6,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9acd5a76-9f07-4525-8c6d-4b5f3ed12aae",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "crested",
   "language": "python",
   "name": "crested"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
